src/Core/Framework/Framework.php

```diff
@@ -166,27 +168,49 @@ private function registerEntityExtensions(
         ExtensionRegistry $registry
     ): void {
         foreach ($registry->getExtensions() as $extension) {
-            /** @var string $class */
-            $class = $extension->getDefinitionClass();
-
-            try {
-                $definition = $definitionRegistry->get($class);
-            } catch (DefinitionNotFoundException) {
-                continue;
-            }
-
-            $definition->addExtension($extension);
-
-            try {
-                $salesChannelDefinition = $salesChannelRegistry->get($class);
-            } catch (DefinitionNotFoundException) {
-                continue;
-            }
-
-            // same definition? do not added extension
-            if ($salesChannelDefinition !== $definition) {
-                $salesChannelDefinition->addExtension($extension);
-            }
+            $this->addExtension($definitionRegistry, $salesChannelRegistry, $extension);
         }
+
+        foreach ($registry->buildBulkExtensions($definitionRegistry) as $extension) {
+            $this->addExtension($definitionRegistry, $salesChannelRegistry, $extension);
+        }
+    }
+
+    private function addExtension(
+        DefinitionInstanceRegistry $definitionRegistry,
+        SalesChannelDefinitionInstanceRegistry $salesChannelRegistry,
+        EntityExtension $extension
+    ): void {
+        try {
+            $definition = $this->getInstance($definitionRegistry, $extension);
+        } catch (DefinitionNotFoundException) {
+            return;
+        }
+
+        $definition->addExtension($extension);
+
+        try {
+            $salesChannelDefinition = $this->getInstance($salesChannelRegistry, $extension);
+        } catch (DefinitionNotFoundException) {
+            return;
+        }
+
+        // same definition? do not added extension
+        if ($salesChannelDefinition !== $definition) {
+            $salesChannelDefinition->addExtension($extension);
+        }
+    }
+
+    private function getInstance(DefinitionInstanceRegistry $registry, EntityExtension $extension): EntityDefinition
+    {
+        if (!empty($extension->getEntityName())) {
```

Comment:
again deprecation handling, i would check the feature flag here and if 6.7 is active only do the look up based on entity name

all the check if empty or fallback to definition should be removed in 6.7